scale_build/image/iso.py

```diff
@@ -0,0 +1,94 @@
+import glob
+import itertools
+import os
+import shutil
+
+from scale_build.config import VERSION
+from scale_build.utils.manifest import get_manifest
+from scale_build.utils.run import run
+from scale_build.utils.paths import CD_DIR, CD_FILES_DIR, CHROOT_BASEDIR, CONF_GRUB, RELEASE_DIR, TMP_DIR
+
+from .manifest import UPDATE_FILE
+from .utils import run_in_chroot
+
+
+def install_iso_packages(iso_logger):
+    run_in_chroot(['apt', 'update'], logger=iso_logger)
+
+    # echo "/dev/disk/by-label/TRUENAS / iso9660 loop 0 0" > ${CHROOT_BASEDIR}/etc/fstab
+    for package in get_manifest()['iso-packages']:
+        run_in_chroot(['apt', 'install', '-y', package], logger=iso_logger)
+
+    os.makedirs(os.path.join(CHROOT_BASEDIR, 'boot/grub'), exist_ok=True)
+    shutil.copy(CONF_GRUB, os.path.join(CHROOT_BASEDIR, 'boot/grub/grub.cfg'))
+
+
+def make_iso_file(iso_logger):
+    for f in glob.glob(os.path.join(RELEASE_DIR, '*.iso*')):
+        os.unlink(f)
+
+    # Set default PW to root
+    run(fr'chroot {CHROOT_BASEDIR} /bin/bash -c "echo -e \"root\nroot\" | passwd root"', logger=iso_logger, shell=True)
+
+    # Create /etc/version
+    with open(os.path.join(CHROOT_BASEDIR, 'etc/version'), 'w') as f:
+        f.write(VERSION)
+
+    # Copy the CD files
+    distutils.dir_util._path_created = {}
+    distutils.dir_util.copy_tree(CD_FILES_DIR, CHROOT_BASEDIR, preserve_symlinks=True)
+
+    # Create the CD assembly dir
+    shutil.rmtree(CD_DIR, ignore_errors=True)
+    os.makedirs(CD_DIR, exist_ok=True)
+
+    # Prune away the fat
+    prune_cd_basedir()
+
+    # Lets make squashfs now
+    tmp_truenas_path = os.path.join(TMP_DIR, 'truenas.squashfs')
+    run(['mksquashfs', CHROOT_BASEDIR, tmp_truenas_path, '-comp', 'xz'], logger=iso_logger)
+    os.makedirs(os.path.join(CD_DIR, 'live'), exist_ok=True)
+    shutil.move(tmp_truenas_path, os.path.join(CD_DIR, 'live/filesystem.squashfs'))
+
+    # Copy over boot and kernel before rolling CD
+    shutil.copytree(os.path.join(CHROOT_BASEDIR, 'boot'), os.path.join(CD_DIR, 'boot'))
+    # Dereference /initrd.img and /vmlinuz so this ISO can be re-written to a FAT32 USB stick using Windows tools
+    shutil.copy(os.path.join(CHROOT_BASEDIR, 'initrd.img'), CD_DIR)
+    shutil.copy(os.path.join(CHROOT_BASEDIR, 'vmlinuz'), CD_DIR)
+    for f in itertools.chain(
+        glob.glob(os.path.join(CD_DIR, 'boot/initrd.img-*')),
+        glob.glob(os.path.join(CD_DIR, 'boot/vmlinuz-*')),
+    ):
+        os.unlink(f)
+
+    shutil.copy(UPDATE_FILE, os.path.join(CD_DIR, 'TrueNAS-SCALE.update'))
+    os.makedirs(os.path.join(CHROOT_BASEDIR, RELEASE_DIR), exist_ok=True)
+    os.makedirs(os.path.join(CHROOT_BASEDIR, CD_DIR), exist_ok=True)
+
+    run(['mount', '--bind', RELEASE_DIR, os.path.join(CHROOT_BASEDIR, RELEASE_DIR)])
+    run(['mount', '--bind', CD_DIR, os.path.join(CHROOT_BASEDIR, CD_DIR)])
+
+    run_in_chroot(['apt-get', 'update'], logger=iso_logger, check=False)
+    run_in_chroot(
```

Comment:
Why `check=False`?

Reply:
Keeping the same behavior as original implementation, checking in shell we have
```
Setting up grub2-common (2.04-16) ...
Setting up grub-pc-bin (2.04-16) ...
Setting up shim-signed:amd64 (1.33+15+1533136590.3beb971-7) ...
Setting up grub-efi-amd64 (2.04-16) ...

Creating config file /etc/default/grub with new version
/usr/sbin/grub-probe: error: cannot find a device for / (is /dev mounted?).
dpkg: error processing package grub-efi-amd64 (--configure):
 installed grub-efi-amd64 package post-installation script subprocess returned error exit status 1
dpkg: dependency problems prevent configuration of grub-efi:
 grub-efi depends on grub-efi-amd64 (= 2.04-16); however:
  Package grub-efi-amd64 is not configured yet.

dpkg: error processing package grub-efi (--configure):
 dependency problems - leaving unconfigured
Processing triggers for libc-bin (2.31-9) ...
Processing triggers for shim-signed:amd64 (1.33+15+1533136590.3beb971-7) ...
Errors were encountered while processing:
 grub-efi-amd64
 grub-efi
E: Sub-process /usr/bin/dpkg returned an error code (1)
root@scale:/# echo $?
100
root@scale:/#
```
So it was always failing but apparently our use case was still being fulfilled.